test/r/result_test.rb

```diff
@@ -5,6 +5,34 @@
 
 module R
   class ResultTest < Minitest::Test
+    describe "R.propagate!" do
+      it "propagates the first try! error" do
+        ok1 = T.let(R.ok(1), R::Result[Integer, String])
+        err1 = T.let(R.err("err1"), R::Result[Integer, String])
+        err2 = T.let(R.err("err2"), R::Result[Integer, String])
+
+        result = R.propagate! do
+          ok1.try!
+          err1.try!
+          err2.try!
+        end
+
+        assert_equal(R.err("err1"), result)
+      end
+
+      it "propagates the last try! ok" do
```

Comment:
FYI, this test is deceiving. I think it's possible to do better, but this code _actually_ just returns the result of the block, whatever that is. Since there's no runtime type checking of generics, I think you could just return an arbitrary `Result` from the block and there wouldn't be any type errors.